### DISCUSSION TITLE
🌼 Spring 🌱 Cleanup 2023

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#5861] [General]
We recently removed support for bitwise operators (`<<<`, `>>>`, and `^^^`). (Don't worry they are still available as functions.)

I think it is nice to re-evaluate "niche" features from time to time.

Hence, I would like to propose (and discuss) the removal of the following features:

- [ ] The `@Unsafe` annotation - barely used and has no real meaning.
- [ ] `Read(r)` and `Write(r)` effects. They are sugar, but they do nothing. I propose to replace them by either just `r` or `Heap(r)`. This does not mean Read/Write could not return, but if they do, they should have a meaning!
- [ ] The `lazy` and `force` constructs.
- [ ] Record selection lambda (bonus points if you even know what they are!)
- [ ] The `opaque` modifier. (What does it do? Does it work?)
- [ ] Disable inlay hints (or at least further reduce when they pop-up).
- [ ] The `--benchmark` and `--test` flags. (I believe they are 99% redundant now that we have the `benchmark` and `test` _commands_.)

Note:  The `rel` / `lat` declarations are already slated for removal. It has just not happened yet.

Do you have any other features you would like to see gone?

**Comment by magnus-madsen** [comment]
@JonathanStarup @mlutze @stephentetley @jaschdoc Any thoughts?

**Comment by mlutze** [comment]
##### Unsafe

Weak disagree. We have functions that should only be used "if you know what you are doing" and library implementers might add more (flixball has a few). Documenting these with a language feature is a much stronger signal than just putting "be careful" in the comments.

##### Lazy/Force

What happens to LazyList?

##### Record selection lambda

Strong agree.

##### Opaque

Strong disagree. The invariants of some types depends on their not being opened. We need more principled use of this feature. I think it currently doesn't work as intended since the namespace change.

##### Benchmark/Test

Strong agree. We really need a more principled approach to our command line interface.

### Others

##### Par

In my opinion, the par-yield is simply more usable and the special treatment of tuples, etc. just complicates everything.

**Reply by magnus-madsen** [reply]
Which par?

**Reply by mlutze** [reply]
I don't like this one:
`par (f(1), 1, 2, x, f(2), x, 2, f(3), x, 2)`

I like this one:
```
        par (
            a <- 1;              // Constant
            b <- c;              // Variable lookup
            d <- (let a = 3; a); // Spawn into thread
            e <- (let b = 4; b); // Last complex expression
            f <- 5;              // Constant
            g <- 6               // Constant
        ) yield a + b + d + e + f + g
```

**Reply by magnus-madsen** [reply]
Roger. Lets put it up for discussion.

**Comment by JonathanStarup** [comment]
I exactly agree with Matt :)

**Comment by magnus-madsen** [comment]
My current sentiment:

- [x] Drop `@Unsafe`
- [ ] Drop `Read(r)` and `Write(r)`.
- [x] Drop record selection lambda.
- [x] Drop `par` tuple.
- [x] Drop `--benchmark` and `--test` flags.
- [ ] Drop `opaque` but bring it back later.
- [ ] Keep `InlayHints`.

**Reply by JonathanStarup** [reply]
A note on `--test`, its nice to run single test files in the compiler with `some/path/text.flix --test`

**Reply by magnus-madsen** [reply]
Yeah, I agree. But it confuses the hell out of our users :/

**Reply by JonathanStarup** [reply]
a sneaky solution would be to remove it from help or add --X

**Reply by magnus-madsen** [reply]
It still has multiple maintenance problem. In any case, its dead now.

**Comment by stephentetley** [comment]
I think we can eliminate the region from `Iterator` which is a nice simplification.

I've re-implemented about 1/4 of the API without the region so far.

**Reply by magnus-madsen** [reply]
Huh? How is this possible? I assume there should be a suspended effect in it? 
Do you mean the region capability? How is that possible? :)

**Reply by magnus-madsen** [reply]
We have a paper in submission where we use `Iterator` as a case study. So I am intrigued (and maybe worried (??)). I thought all the complexity was inherent (necessary).

**Comment by stephentetley** [comment]
Simplifying to this type:

~~~

pub enum Iterator[ef: Bool, a: Type] {
    case Iterator(Unit -> Iterator.Step[a] \ ef)
}

~~~

It might be that I can't write `flatMap` for this type so it might not be a worthwhile replacement. I'll be working on `flatMap` this evening.

**Reply by magnus-madsen** [reply]
This will probably not work... Because we do not have sub-effecting and the region of the iterator could be different from that of the effect you map with. I think with user-defined effects this will become more clear.

But, I would still applaud an attempt at flatMap with only this definition.

**Reply by magnus-madsen** [reply]
Hmm. I could be wrong. Maybe disregard everything I said...

**Comment by stephentetley** [comment]
Oh dear, the Iterator formulation above makes easy things simpler but makes hard things (e.g. iterators implemented with mutable refs like `flatMap`) much harder.

One thing it did do was remind me of the "double thunk" style of iterators - this style essentially wraps the outer mutable initialization of the iterators we have already in another thunk. This looks like it has the very useful property of catching more of the effects inside the type constructor rather than having them "float" to the right hand side of the type signature.

For instance the original signature for `repeat` was:

> pub def repeat(rc: Region[r], n: Int32, x: a): Iterator[a, r, r] \ Write(r)

But now it is: 

> pub def repeat(rc: Region[r], n: Int32, x: a): Iterator3[a, r, r]

The initialization and updating of the ref cell are captured entirely in the constructor.


~~~


pub enum Iterator3[a: Type, ef: Bool, r: Region] {
    case Iterator3(Region[r], Iterator3.GenBody[a, ef, r])
}

mod Iterator3 {
    use Step.{Ans, Skip, Done}
    use GenBody.{GenBody}


    ///
    /// Represent the "steps" of an iterator.
    ///
    /// This is an expanded version of the `Option` type.
    /// `Ans` is the answer at an intermediate stage of iteration.
    /// `Done` signals that iteration is complete.
    /// `Skip` is essentially an implementation detail. It allows us
    /// to define `filter` and similar functions that "wither" Iterators
    /// without recursively running through the Iterator.
    ///
    pub enum Step[a] {
        case Ans(a)
        case Skip
        case Done
    }

    pub enum GenBody[a: Type, ef: Bool, r: Region](
        Unit -> (Unit -> Step[a] \ { ef, r }) \ r
    )

    ///
    /// Returns an iterator over an iterable with the element `x` repeated `n` times.
    ///
    /// Returns an empty iterator if `n < 0`.
    ///
    pub def repeat(rc: Region[r], n: Int32, x: a): Iterator3[a, r, r] =
        Iterator3(rc, repeatBody(rc, n, x))

    pub def repeatBody(rc: Region[r], n: Int32, x: a): GenBody[a, ef or r, r] =
        GenBody(_ ->
            let ix = ref n @ rc;
            let iterF = () -> {
                let i = deref ix;
                if (i < 1)
                    Done
                else {
                    ix := i-1;
                    Ans(x)
                }
            };
            iterF
        )
    ///
    /// Applies `f` to every element of `iter`.
    ///
    /// Consumes the entire iterator.
    ///
    pub def forEach(f: a -> Unit \ ef2, iter: Iterator3[a, ef1, r]): Unit \ { ef1, ef2, r } =
        let Iterator3(_, GenBody(genIter)) = iter;
        let iterF = genIter();
        def loop() = match iterF() {
            case Skip   => loop()
            case Ans(a) => f(a); loop()
            case Done   => ()
        };
        loop()


    pub def test01(): Unit \ IO = region rc {
        let iter1 = repeat(rc, 5, 'o');
        forEach(println, iter1)
    }
}

~~~

**Comment by stephentetley** [comment]
I've written a full replacement `Iterator` module including `flatMap` and `intercalate` in the above style but I didn't realize `next` doesn't work and goes into an infinite loop until I was in a position to run the existing tests. The types were significantly nicer but as it didn't work that doesn't count.

**Comment by stephentetley** [comment]
Back to non-Iterator things, now that we don't have the `Scoped` type class should type constructors of mutable collections have there arguments flipped, e.g. :

> pub enum MutDeque[r: Region, a: Type]

**Reply by mlutze** [reply]
I am strongly in favor of switching. With associated types we can always bring back `Scoped` if needed
